rr_iarrc/src/finish_line_watcher/finish_line_watcher.cpp

```diff
@@ -78,23 +84,122 @@ void ImageCB(const sensor_msgs::ImageConstPtr& msg) {
+                                         std::end(contours.at(i)));
+            color = cv::Scalar(0, 255, 0);
+        }
+
+        // Debug draw
+        cv::drawContours(debugDrawing, contours, i, color, 2, 8);
+    }
+
+    bool detected = false;
+    cv::RotatedRect fitRect;
+    cv::Size2f size;
+    float angle;
+
+    if (filteredContourPoints.size() > 0) {
+        // Fit a rect to the points and detect based on angle and width
+        fitRect = cv::minAreaRect(filteredContourPoints);
+        size = fitRect.size;
+        angle = std::abs(std::abs(fitRect.angle) - 90);
+        // Check the detection criteria (must look at both width and height because OpenCV is inconsistent about how it
+        // assigns them)
+        detected = angle < angle_cutoff && (size.width > width_cutoff || size.height > width_cutoff);
+
+        Point2f rectPoints[4];
+        fitRect.points(rectPoints);
+
+        for (int i = 0; i < 4; i++)
+            cv::line(debugDrawing, rectPoints[i], rectPoints[(i + 1) % 4], cv::Scalar(255, 0, 0), 2, 12);
+    }
+
+    // Quick count of pixels as a final sanity check
+    auto count = cv::countNonZero(frame);
+    detected = detected && count > count_thresh;
+    auto incrementCrossNum = false;
+
+    if (publish_when_detected) {
+        // Publish when first detected but don't keep on increasing number of crosses, so make use of states again
+        if (detected && state == LOW && cooldown == 0) {
+            incrementCrossNum = true;
+            state = HIGH;
+            cooldown = cooldown_value;
+        } else if (state == HIGH && !detected) {
+            state = LOW;
+        }
 
-    auto width = getWidth(frame);
+    } else {
+        // When the line is first detected, set to a high state then when we stop detecting (we've crossed), report that
+        if (state == LOW && detected && cooldown == 0) {
+            state = HIGH;
+        } else if (state == HIGH && !detected) {
+            state = LOW;
+            incrementCrossNum = true;
+        }
+    }
 
-    if (state == LOW && count > 2000 && width > 700) {
-        state = HIGH;
-    } else if (state == HIGH && count < 2000) {
-        state = LOW;
+    if (incrementCrossNum) {
         number_of_crosses++;
+        cooldown = cooldown_value;
         ROS_INFO_STREAM("Finish line crossed: " << to_string(number_of_crosses));
     }
 
-    // if(count > 2000 && width > 600) {
-    //     number_of_crosses = 1;
-    //     ROS_INFO_STREAM("Finish line crossed: " <<
-    //     to_string(number_of_crosses));
-    // }
+    if (cooldown > 0)
+        cooldown--;
+
+    // Publish stuff
+    std_msgs::Int8 intmsg;
+    intmsg.data = number_of_crosses;
+
+    if (debug_pub.getNumSubscribers() > 0) {
+        // Draw some debug info
+        std::string angle_str = std::string("Angle: ") + std::to_string(angle);
+        cv::putText(debugDrawing, angle_str, cv::Point(5, 100), cv::FONT_HERSHEY_SIMPLEX, 2, cv::Scalar(0, 143, 143),
```

Comment:
Probably should check that `fitRect` is initialized when you are printing angle and stuff